rxjs-1/carsLeaderBoard.js

```diff
@@ -0,0 +1,144 @@
+  return {
+    time: time,
+    carName: carName,
+    xLocation: xLocation,
+    speed: 0
+  }
+}
+
+// we prepare the board raw data in order to complete the computation
+const prepareDataProcessBoard = (board) => {
+  let pos = CAR_NUMBER;
+  // we order by inverted position
+  // in order to simplify the calcul of the gap
+  return sortTableObject(board, 'xLocation').map((car) => {
+      const ret = {
+        ...car,
+        position: pos,
+        leaderGapDistance: 0,
+        leaderGapTime: 0,
+        speed: car.speed
+
+      };
+      pos--;
+      return ret;
+    });
+  return
+}
+
+// finalize the data of the board comptuation
+const processBoard = (board) => {
+  const retBoard = prepareDataProcessBoard(board);
+  // we process with the calculation of the inverted list
+  retBoard.forEach((car, index) => {
+    if (index < CAR_NUMBER) {
+      const nextCar = retBoard[index+1];
+      // just to be sure ...
+      if (nextCar) {
+        car.leaderGapDistance = nextCar.xLocation - car.xLocation;
+      }
+    }
+
+  });
+  return (retBoard)
+}
+
+// calcul just the gap leader time
+// last computation before send the data to the view
+const processSpeed = (board) => {
+  // we process with the calculation of the inverted list
+  board.forEach((car, index) => {
+    if (index < CAR_NUMBER) {
+      const nextCar = board[index+1];
+      // just to be sure ...
+      if (nextCar) {
+        // here the mutliplier transform the value from per seconds to per milliseconds
+        car.leaderGapTime = ((car.leaderGapDistance / car.speed) * 1000);
+      }
+    }
+
+  });
+  return (board)
+}
+
+const mergeData = ([leaderboard, ...speeds]) => {
+  finalBoard.forEach((board, i) => {
+    // get the speed in the order from the stream
+    // at this moment our finalboard is in the same position of our stream
+    board.speed = speeds[i];
+  })
+  // sort our board by postion to be usable in the view
+  finalBoard = sortTableObject(processSpeed(finalBoard), 'position');
+  return finalBoard;
+}
+
+
+const leaderboardData = ({time, carName, xLocation}) => {
+  // avoid duplicate data in test exemple
+  if (!carExistInBoard(currentLeaderBoardData, carName)) {
+    // push in the current data board the information of the car present
+    currentLeaderBoardData.push(baseCarObject(time, carName, xLocation));
+  }
+
+  // Limit to the number of cars in the exemple
+  if (currentLeaderBoardData.length >= CAR_NUMBER) {
+    // process the current board raw data
+    finalBoard = processBoard(currentLeaderBoardData);
+    currentLeaderBoardData = []
+  }
+}
+
+const leaderBoardObservable = (race) => {
+  const observable = new Observable(subscriber => {
+
+    race.on('data', ({time, carName, xLocation}) => {
+      subscriber.next({time, carName, xLocation});
+    });
+    race.on('end', () => {
+      subscriber.complete();
+    });
+  });
+  return observable
+    .pipe(map(leaderboardData))
```

Comment:
This generates a side-effect (it writes to `currentLeaderBoardData` and `finalBoard`), but it shouldn't